### PR TITLE
Instant Launch Tables, Constraints, and Conversion

### PR DESCRIPTION
**Pull request by johnworth**
Contains the initial implementation of the instant launch tables, constraints, and conversion. 

The mappings of info-types and file extensions to instant-launches are contained in JSONB fields.

### CODE REVIEW

File: src/main/conversions/v2.35.0/c2_35_0_2020110601.clj [hunk 1]
```diff
@@ -0,0 +1,28 @@
+(ns facepalm.c2-35-0-2020110601
+  (:use [kameleon.sql-reader :only [load-sql-file]]))
+
+(def ^:private version
+  "The destination database version"
+  "2.35.0:20201106.01")
```
Comment (slr71) [inline src/main/conversions/v2.35.0/c2_35_0_2020110601.clj:6]: I'll need to look this over more thoroughly this afternoon, but I did want to quickly mention this. There should be a corresponding line in `src/main/data/999_version.sql` to ensure that the correct version is applied in `init` mode.

Comment (johnworth) [reply src/main/conversions/v2.35.0/c2_35_0_2020110601.clj:6]: Oops, sorry. I always forget that. Thanks for pointing it out.

Comment (johnworth) [reply src/main/conversions/v2.35.0/c2_35_0_2020110601.clj:6]: Should be fixed now.

File: src/main/tables/107_default_instant_launches.sql [hunk 2]
```diff
@@ -0,0 +1,13 @@
+SET search_path = public, pg_catalog;
+
+CREATE TABLE IF NOT EXISTS default_instant_launches (
+    id UUID UNIQUE NOT NULL DEFAULT uuid_generate_v1(),
+
+    version SERIAL NOT NULL,
+
+    instant_launches JSONB NOT NULL DEFAULT '{}'::JSONB,
```
Comment (slr71) [inline src/main/tables/107_default_instant_launches.sql:40]: It might be useful to have a comment either in the initialization files or in the database itself describing how this column is used to map info type names and glob patterns to compatible instant launch IDs.

File: src/main/tables/108_user_instant_launches.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+SET search_path = public, pg_catalog;
+
+CREATE TABLE IF NOT EXISTS user_instant_launches (
+    id UUID UNIQUE NOT NULL DEFAULT uuid_generate_v1(),
+
+    user_id UUID UNIQUE NOT NULL,
+
+    instant_launches JSONB NOT NULL DEFAULT '{}'::JSONB,
```
Comment (slr71) [inline src/main/tables/108_user_instant_launches.sql:41]: A comment might be useful on this column too.

File: src/main/tables/108_user_instant_launches.sql [hunk 4]
```diff
@@ -0,0 +1,15 @@
+SET search_path = public, pg_catalog;
+
+CREATE TABLE IF NOT EXISTS user_instant_launches (
+    id UUID UNIQUE NOT NULL DEFAULT uuid_generate_v1(),
+
+    user_id UUID UNIQUE NOT NULL,
```
Comment (slr71) [inline src/main/tables/108_user_instant_launches.sql:12]: Since we have a version on default instant launch settings, we might need a version of user-specific instant launch settings as well. Also, a uniqueness constraint on the `user_id` and `version` columns might be useful here.

File: src/main/constraints/107_default_instant_launches.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+SET search_path = public, pg_catalog;
+
+ALTER TABLE default_instant_launches
+    ADD CONSTRAINT default_instant_launches_added_by_fkey
+    FOREIGN KEY (added_by)
+    REFERENCES users(id);
+
+ALTER TABLE default_instant_launches
+    ADD CONSTRAINT default_instant_launches_id_version_unique
+    UNIQUE (id, version);
```
Comment (slr71) [inline src/main/constraints/107_default_instant_launches.sql]: If I'm understanding things correctly, it seems to me that this uniqueness constraint should be applied only to the `version` column.

**Review by slr71 — COMMENTED** [review]
Looks good. I had a few suggestions, but nothing major.

File: src/main/tables/107_default_instant_launches.sql [hunk 6]
```diff
@@ -0,0 +1,13 @@
+SET search_path = public, pg_catalog;
+
+CREATE TABLE IF NOT EXISTS default_instant_launches (
+    id UUID UNIQUE NOT NULL DEFAULT uuid_generate_v1(),
+
+    version SERIAL NOT NULL,
```
Comment (ianmcorvidae) [inline src/main/tables/107_default_instant_launches.sql]: I'm not sure if I'm missing something, but it seems like these columns duplicate each other. The version here being set to `SERIAL` suggests to me that this field will be globally unique, rather than being a version within some given set. Otherwise, we'd need to be manually inserting things that ignore the sequence (i.e., don't use the default value), but I'm not sure what that would be useful for.

Basically, I'm curious if there's a reason to have the `id` column rather than using `version` as the primary key.

As a side note, `SERIAL` already implies `NOT NULL`, re [docs](https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-SERIAL) -- basically it works out to `integer NOT NULL DEFAULT nextval('the_sequence_created_automatically_by_postgres_for_this')`

File: src/main/constraints/107_default_instant_launches.sql (hunk 5 above)
Comment (ianmcorvidae) [reply src/main/constraints/107_default_instant_launches.sql]: Yeah, I think this won't do much as is, since `id` is already unique. It will create an index on `(id, version)`, but particularly with the unique column first I'm not sure it's ary more useful than the primary key index on `id`.

**Review by ianmcorvidae — COMMENTED** [review]
Mostly I have one question about how the default instant launch versions are supposed to work (the current table makes it seem weird) and otherwise mostly just agree with Sarah's review comments, so I won't duplicate all of them :)

File: src/main/tables/107_default_instant_launches.sql (hunk 2 above)
Comment (johnworth) [reply src/main/tables/107_default_instant_launches.sql:40]: Comments added.

File: src/main/tables/108_user_instant_launches.sql (hunk 3 above)
Comment (johnworth) [reply src/main/tables/108_user_instant_launches.sql:41]: Comment added.

File: src/main/constraints/107_default_instant_launches.sql (hunk 5 above)
Comment (johnworth) [reply src/main/constraints/107_default_instant_launches.sql]: I removed the id field from the constraint.

File: src/main/tables/107_default_instant_launches.sql (hunk 6 above)
Comment (johnworth) [reply src/main/tables/107_default_instant_launches.sql]: I added a comment to the file to this effect, but the intention is that we can maintain a history of edits to the default instant launches so that we can make edits safely without having to parse or modify primary keys in the database.

Comment (johnworth) [reply src/main/tables/107_default_instant_launches.sql]: Actually, my comment is incorrect. Fixing now. Sorry, caffeine hasn't kicked in fully yet.

Comment (johnworth) [reply src/main/tables/107_default_instant_launches.sql]: Okay, now it's fixed.

File: src/main/tables/108_user_instant_launches.sql (hunk 4 above)
Comment (johnworth) [reply src/main/tables/108_user_instant_launches.sql:12]: Added, thanks.

File: src/main/constraints/108_user_instant_launches.sql [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    REFERENCES users(id) ON DELETE CASCADE;
+
+ALTER TABLE user_instant_launches
+    ADD CONSTRAINT user_instant_launches_added_by_fkey
+    FOREIGN KEY (added_by)
+    REFERENCES users(id);
+
+ALTER TABLE user_instant_launches
+    ADD CONSTRAINT user_instant_launches_version_unique
+    UNIQUE (version);
```
Comment (slr71) [inline src/main/constraints/108_user_instant_launches.sql]: This constraint should probably be on both the `version` and `user_id` columns.

File: src/main/tables/107_default_instant_launches.sql [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    -- {
+    --     "*.py" : {
+    --         "kind" : "pattern",
+    --         "default" : "<instant_launch_uuid>",
+    --         "compatible: [
+    --             "<instant_launch_uuid>",
+    --         ],
+    --     },
+
+    --     "<info_type_uuid>" : {
```
Comment (slr71) [inline src/main/tables/107_default_instant_launches.sql]: I think that the info type name will have to be used for this. The `info_type` table in the DE database is out of date and eventually either needs to be updated or removed.

File: src/main/tables/108_user_instant_launches.sql [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    -- {
+    --     "*.py" : {
+    --         "kind" : "pattern",
+    --         "default" : "<instant_launch_uuid>",
+    --         "compatible: [
+    --             "<instant_launch_uuid>",
+    --         ],
+    --     },
+
+    --     "<info_type_uuid>" : {
```
Comment (slr71) [inline src/main/tables/108_user_instant_launches.sql]: Another case where the UUID can't be used.

**Review by slr71 — APPROVED** [review]
I caught a couple minor things, but nothing major. :shipit: whenever you're ready.

File: src/main/tables/107_default_instant_launches.sql [hunk 10]
```diff
@@ -0,0 +1,45 @@
+SET search_path = public, pg_catalog;
+
+CREATE TABLE IF NOT EXISTS default_instant_launches (
+    id UUID UNIQUE NOT NULL DEFAULT uuid_generate_v1(),
+
+    -- Indicates the format version of the instant launch
+    -- JSON. Should help us put together conversions between
+    -- format versions without breaking currently running
+    -- services.
+    version SERIAL,
```
Comment (ianmcorvidae) [inline src/main/tables/107_default_instant_launches.sql:10]: Ah, cool, this makes sense. `version` being for json-data-format versions rather than a history of changes to the set of defaults makes a lot more sense, since I guess basically this means "run `UPDATE` on the JSON column if the format isn't changing, otherwise create a new one with a new version number"

File: src/main/tables/107_default_instant_launches.sql (hunk 8 above)
Comment (ianmcorvidae) [reply src/main/tables/107_default_instant_launches.sql]: I agree -- the names are usable keys anyway, since that's what we store in AVUs, so probably we can just remove any extra surrogate keys in our database, is my feeling.

**Review by ianmcorvidae — APPROVED** [review]
Looks reasonable to me. I think Sarah's right about the one UNIQUE constraint also, but that's the only issue I see, so I'm marking it approved.

File: src/main/tables/107_default_instant_launches.sql (hunk 8 above)
Comment (slr71) [reply src/main/tables/107_default_instant_launches.sql]: I like the idea of removing the surrogate keys in the database.

File: src/main/constraints/108_user_instant_launches.sql (hunk 7 above)
Comment (johnworth) [reply src/main/constraints/108_user_instant_launches.sql]: Fixed. Thanks.

File: src/main/tables/107_default_instant_launches.sql (hunk 8 above)
Comment (johnworth) [reply src/main/tables/107_default_instant_launches.sql]: Fixed the comments.

File: src/main/tables/108_user_instant_launches.sql (hunk 9 above)
Comment (johnworth) [reply src/main/tables/108_user_instant_launches.sql]: Fixed the comment.
